In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from spicy import stats

plt.rcParams['figure.figsize'] = (10,5)
plt.rcParams['axes.grid'] = True

In [3]:
df = pd.read_csv(r'dataset\diminos_data.csv')
df

order_id      order_placed_at          order_delivered_at
0       1523111  2023-03-01 00:00:59  2023-03-01 00:18:07.443132
1       1523112  2023-03-01 00:03:59  2023-03-01 00:19:34.925241
2       1523113  2023-03-01 00:07:22  2023-03-01 00:22:28.291385
3       1523114  2023-03-01 00:07:47  2023-03-01 00:46:19.019399
4       1523115  2023-03-01 00:09:03  2023-03-01 00:25:13.619056
...         ...                  ...                         ...
14995   1538106  2023-03-27 23:37:05  2023-03-27 23:52:37.409378
14996   1538107  2023-03-27 23:47:38  2023-03-28 00:04:22.672912
14997   1538108  2023-03-27 23:50:16  2023-03-28 00:05:40.676238
14998   1538109  2023-03-27 23:52:44  2023-03-28 00:08:41.810358
14999   1538110  2023-03-27 23:58:20  2023-03-28 00:13:42.499311

[15000 rows x 3 columns]

In [4]:
df.head()

order_id      order_placed_at          order_delivered_at
0   1523111  2023-03-01 00:00:59  2023-03-01 00:18:07.443132
1   1523112  2023-03-01 00:03:59  2023-03-01 00:19:34.925241
2   1523113  2023-03-01 00:07:22  2023-03-01 00:22:28.291385
3   1523114  2023-03-01 00:07:47  2023-03-01 00:46:19.019399
4   1523115  2023-03-01 00:09:03  2023-03-01 00:25:13.619056

In [5]:
df.shape

(15000, 3)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            15000 non-null  int64 
 1   order_placed_at     15000 non-null  object
 2   order_delivered_at  15000 non-null  object
dtypes: int64(1), object(2)
memory usage: 351.7+ KB


In [7]:
df.columns

Index(['order_id', 'order_placed_at', 'order_delivered_at'], dtype='object')

In [8]:
df.describe()

order_id
count  1.500000e+04
mean   1.530610e+06
std    4.330271e+03
min    1.523111e+06
25%    1.526861e+06
50%    1.530610e+06
75%    1.534360e+06
max    1.538110e+06

In [9]:
# date time conversion 
df['order_placed_at'] = pd.to_datetime(df['order_placed_at'])
df['order_delivered_at'] = pd.to_datetime(df['order_delivered_at'])
df

order_id     order_placed_at         order_delivered_at
0       1523111 2023-03-01 00:00:59 2023-03-01 00:18:07.443132
1       1523112 2023-03-01 00:03:59 2023-03-01 00:19:34.925241
2       1523113 2023-03-01 00:07:22 2023-03-01 00:22:28.291385
3       1523114 2023-03-01 00:07:47 2023-03-01 00:46:19.019399
4       1523115 2023-03-01 00:09:03 2023-03-01 00:25:13.619056
...         ...                 ...                        ...
14995   1538106 2023-03-27 23:37:05 2023-03-27 23:52:37.409378
14996   1538107 2023-03-27 23:47:38 2023-03-28 00:04:22.672912
14997   1538108 2023-03-27 23:50:16 2023-03-28 00:05:40.676238
14998   1538109 2023-03-27 23:52:44 2023-03-28 00:08:41.810358
14999   1538110 2023-03-27 23:58:20 2023-03-28 00:13:42.499311

[15000 rows x 3 columns]

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   order_id            15000 non-null  int64         
 1   order_placed_at     15000 non-null  datetime64[ns]
 2   order_delivered_at  15000 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1)
memory usage: 351.7 KB


In [11]:
# Check for missing values
df.isnull().sum()

order_id              0
order_placed_at       0
order_delivered_at    0
dtype: int64

In [12]:
df['delivery_time_mins'] = (df['order_delivered_at'] - df['order_placed_at']).dt.total_seconds() / 60
df

order_id     order_placed_at         order_delivered_at  \
0       1523111 2023-03-01 00:00:59 2023-03-01 00:18:07.443132   
1       1523112 2023-03-01 00:03:59 2023-03-01 00:19:34.925241   
2       1523113 2023-03-01 00:07:22 2023-03-01 00:22:28.291385   
3       1523114 2023-03-01 00:07:47 2023-03-01 00:46:19.019399   
4       1523115 2023-03-01 00:09:03 2023-03-01 00:25:13.619056   
...         ...                 ...                        ...   
14995   1538106 2023-03-27 23:37:05 2023-03-27 23:52:37.409378   
14996   1538107 2023-03-27 23:47:38 2023-03-28 00:04:22.672912   
14997   1538108 2023-03-27 23:50:16 2023-03-28 00:05:40.676238   
14998   1538109 2023-03-27 23:52:44 2023-03-28 00:08:41.810358   
14999   1538110 2023-03-27 23:58:20 2023-03-28 00:13:42.499311   

       delivery_time_mins  
0               17.140719  
1               15.598754  
2               15.104856  
3               38.533657  
4               16.176984  
...                   ...  
14995           15.540156  
14996           16.744549  
14997           15.411271  
14998           15.963506  
14999           15.374989  

[15000 rows x 4 columns]

In [13]:
df['delivery_time_mins'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 15000 entries, 0 to 14999
Series name: delivery_time_mins
Non-Null Count  Dtype  
--------------  -----  
15000 non-null  float64
dtypes: float64(1)
memory usage: 117.3 KB


In [14]:
df['delivery_time_mins'].describe()

count    15000.000000
mean        20.499389
std         96.160362
min         15.000010
25%         15.274826
50%         15.797986
75%         17.279661
max       7299.831375
Name: delivery_time_mins, dtype: float64

In [15]:
df[df['delivery_time_mins'] <= 0]

Empty DataFrame
Columns: [order_id, order_placed_at, order_delivered_at, delivery_time_mins]
Index: []

In [16]:
df[df['delivery_time_mins'] >= 60]

order_id     order_placed_at         order_delivered_at  \
24      1523135 2023-03-01 01:11:53 2023-03-01 05:07:54.563978   
117     1523228 2023-03-01 04:59:04 2023-03-01 18:06:52.046891   
408     1523519 2023-03-01 18:05:17 2023-03-01 19:46:42.833697   
460     1523571 2023-03-01 20:29:53 2023-03-02 01:17:46.815054   
565     1523676 2023-03-02 00:36:51 2023-03-02 02:30:07.988260   
...         ...                 ...                        ...   
14028   1537139 2023-03-26 03:57:24 2023-03-26 05:31:26.392441   
14155   1537266 2023-03-26 09:54:24 2023-03-29 02:42:50.645252   
14181   1537292 2023-03-26 11:29:46 2023-03-26 12:46:36.794410   
14593   1537704 2023-03-27 05:57:43 2023-03-27 07:01:35.117442   
14841   1537952 2023-03-27 16:42:37 2023-03-27 17:43:13.294744   

       delivery_time_mins  
24             236.026066  
117            787.800782  
408            101.430562  
460            287.896918  
565            113.283138  
...                   ...  
14028           94.039874  
14155         3888.444088  
14181           76.846574  
14593           63.868624  
14841           60.604912  

[170 rows x 4 columns]

In [17]:
df[df['delivery_time_mins'] >= 90]

order_id     order_placed_at         order_delivered_at  \
24      1523135 2023-03-01 01:11:53 2023-03-01 05:07:54.563978   
117     1523228 2023-03-01 04:59:04 2023-03-01 18:06:52.046891   
408     1523519 2023-03-01 18:05:17 2023-03-01 19:46:42.833697   
460     1523571 2023-03-01 20:29:53 2023-03-02 01:17:46.815054   
565     1523676 2023-03-02 00:36:51 2023-03-02 02:30:07.988260   
...         ...                 ...                        ...   
13744   1536855 2023-03-25 16:25:15 2023-03-25 18:59:25.756029   
13805   1536916 2023-03-25 19:02:45 2023-03-25 20:53:15.176758   
13828   1536939 2023-03-25 19:56:43 2023-03-25 23:02:44.977694   
14028   1537139 2023-03-26 03:57:24 2023-03-26 05:31:26.392441   
14155   1537266 2023-03-26 09:54:24 2023-03-29 02:42:50.645252   

       delivery_time_mins  
24             236.026066  
117            787.800782  
408            101.430562  
460            287.896918  
565            113.283138  
...                   ...  
13744          154.179267  
13805          110.502946  
13828          186.032962  
14028           94.039874  
14155         3888.444088  

[97 rows x 4 columns]

In [18]:
df[df['delivery_time_mins'] >= 120]

order_id     order_placed_at         order_delivered_at  \
24      1523135 2023-03-01 01:11:53 2023-03-01 05:07:54.563978   
117     1523228 2023-03-01 04:59:04 2023-03-01 18:06:52.046891   
460     1523571 2023-03-01 20:29:53 2023-03-02 01:17:46.815054   
578     1523689 2023-03-02 01:26:48 2023-03-02 03:57:34.693308   
618     1523729 2023-03-02 03:24:46 2023-03-02 07:32:35.703725   
...         ...                 ...                        ...   
12813   1535924 2023-03-24 00:14:08 2023-03-24 03:56:25.950915   
13634   1536745 2023-03-25 11:33:36 2023-03-25 13:37:23.367168   
13744   1536855 2023-03-25 16:25:15 2023-03-25 18:59:25.756029   
13828   1536939 2023-03-25 19:56:43 2023-03-25 23:02:44.977694   
14155   1537266 2023-03-26 09:54:24 2023-03-29 02:42:50.645252   

       delivery_time_mins  
24             236.026066  
117            787.800782  
460            287.896918  
578            150.778222  
618            247.828395  
...                   ...  
12813          222.299182  
13634          123.789453  
13744          154.179267  
13828          186.032962  
14155         3888.444088  

[69 rows x 4 columns]

In [19]:
df[df['delivery_time_mins'] >= 500]

order_id     order_placed_at         order_delivered_at  \
117     1523228 2023-03-01 04:59:04 2023-03-01 18:06:52.046891   
1361    1524472 2023-03-03 10:04:13 2023-03-04 03:51:02.368715   
1910    1525021 2023-03-04 10:41:17 2023-03-04 23:20:15.509579   
2741    1525852 2023-03-05 22:16:39 2023-03-06 07:23:17.239715   
4165    1527276 2023-03-08 09:46:43 2023-03-12 11:34:09.085175   
5109    1528220 2023-03-10 01:04:58 2023-03-10 14:13:37.181329   
5140    1528251 2023-03-10 02:50:34 2023-03-10 11:58:13.378454   
5499    1528610 2023-03-10 17:06:22 2023-03-15 18:46:11.882496   
6311    1529422 2023-03-12 03:47:59 2023-03-12 16:18:04.919941   
10007   1533118 2023-03-19 00:56:34 2023-03-21 00:59:41.454974   
10225   1533336 2023-03-19 11:07:32 2023-03-22 13:53:25.472592   
10712   1533823 2023-03-20 08:13:11 2023-03-20 18:55:51.187897   
11944   1535055 2023-03-22 12:50:53 2023-03-23 08:03:40.654492   
14155   1537266 2023-03-26 09:54:24 2023-03-29 02:42:50.645252   

       delivery_time_mins  
117            787.800782  
1361          1066.822812  
1910           758.975160  
2741           546.637329  
4165          5867.434753  
5109           788.653022  
5140           547.656308  
5499          7299.831375  
6311           750.098666  
10007         2883.124250  
10225         4485.891210  
10712          642.669798  
11944         1152.794242  
14155         3888.444088

In [20]:
df[df['delivery_time_mins'] >= 1000]

order_id     order_placed_at         order_delivered_at  \
1361    1524472 2023-03-03 10:04:13 2023-03-04 03:51:02.368715   
4165    1527276 2023-03-08 09:46:43 2023-03-12 11:34:09.085175   
5499    1528610 2023-03-10 17:06:22 2023-03-15 18:46:11.882496   
10007   1533118 2023-03-19 00:56:34 2023-03-21 00:59:41.454974   
10225   1533336 2023-03-19 11:07:32 2023-03-22 13:53:25.472592   
11944   1535055 2023-03-22 12:50:53 2023-03-23 08:03:40.654492   
14155   1537266 2023-03-26 09:54:24 2023-03-29 02:42:50.645252   

       delivery_time_mins  
1361          1066.822812  
4165          5867.434753  
5499          7299.831375  
10007         2883.124250  
10225         4485.891210  
11944         1152.794242  
14155         3888.444088

In [21]:
def interpret_delivery_time(x):
    if x <= 31:
        return "On Time"
    elif x <= 60:
        return "Late"
    elif x <= 90:
        return "Very Late"
    elif x <= 120:
        return "Extreme Delay"
    else:
        return "System / Closure Artifact"

df['delivery_time_category'] = df['delivery_time_mins'].apply(interpret_delivery_time)

In [22]:
df

order_id     order_placed_at         order_delivered_at  \
0       1523111 2023-03-01 00:00:59 2023-03-01 00:18:07.443132   
1       1523112 2023-03-01 00:03:59 2023-03-01 00:19:34.925241   
2       1523113 2023-03-01 00:07:22 2023-03-01 00:22:28.291385   
3       1523114 2023-03-01 00:07:47 2023-03-01 00:46:19.019399   
4       1523115 2023-03-01 00:09:03 2023-03-01 00:25:13.619056   
...         ...                 ...                        ...   
14995   1538106 2023-03-27 23:37:05 2023-03-27 23:52:37.409378   
14996   1538107 2023-03-27 23:47:38 2023-03-28 00:04:22.672912   
14997   1538108 2023-03-27 23:50:16 2023-03-28 00:05:40.676238   
14998   1538109 2023-03-27 23:52:44 2023-03-28 00:08:41.810358   
14999   1538110 2023-03-27 23:58:20 2023-03-28 00:13:42.499311   

       delivery_time_mins delivery_time_category  
0               17.140719                On Time  
1               15.598754                On Time  
2               15.104856                On Time  
3               38.533657                   Late  
4               16.176984                On Time  
...                   ...                    ...  
14995           15.540156                On Time  
14996           16.744549                On Time  
14997           15.411271                On Time  
14998           15.963506                On Time  
14999           15.374989                On Time  

[15000 rows x 5 columns]

In [23]:
df['delivery_time_category'].value_counts()

delivery_time_category
On Time                      14443
Late                           387
Very Late                       73
System / Closure Artifact       69
Extreme Delay                   28
Name: count, dtype: int64

In [24]:
df['delivery_time_category'].value_counts(normalize=True) * 100

delivery_time_category
On Time                      96.286667
Late                          2.580000
Very Late                     0.486667
System / Closure Artifact     0.460000
Extreme Delay                 0.186667
Name: proportion, dtype: float64

**Delivery time was calculated in minutes to align with Domino’s SLA definition. While most orders were delivered within the expected operational range, a small number exhibited extremely long durations extending several hours. Such values are not operationally realistic for a quick-service delivery model and likely represent delayed order closure or system-level artifacts rather than actual delivery performance. These records were therefore excluded from 95th percentile SLA evaluation to avoid distortion of the metric. All other late and extreme but plausible delays were retained to accurately capture operational risk.**

In [25]:
df_all = df.copy()

In [26]:
df_SLA = df[df['delivery_time_category'] != 'System / Closure Artifact']

In [27]:
df_SLA.shape

(14931, 5)

In [29]:
p95_dt = np.percentile(df_SLA['delivery_time_mins'],95)
p95_dt

np.float64(26.247372241666667)

In [30]:
sla_limit = 31

if p95_dt <= sla_limit :
    print('SLA Met : delivery time is well within 95th percentile')
else :
    print('SLA Breached : 95th percentile exceeds 31 mins')

SLA Met : delivery time is well within 95th percentile


In [31]:
p95_raw = np.percentile(df_all['delivery_time_mins'], 95)
p95_adjusted = p95_dt

p95_raw, p95_adjusted

(np.float64(27.261043996666658), np.float64(26.247372241666667))

**Based on the analysis of delivery performance, the store currently meets Domino’s 95th percentile SLA requirement, with the 95th percentile delivery time remaining well below the 31-minute threshold. As a result, there is no immediate SLA breach or associated financial burden from free-order penalties at this time. Overall operational performance is strong; however, the following recommendations are proposed as preventive measures to ensure continued SLA compliance under future demand fluctuations.**

# Operational Recomendations 

- The current operational environment should be preserved since the current performance is excellent, with over 96% of orders being fulfilled within the SLA.

- Introduce new delivery rider capacity only during peak-risk hours identified through 95th percentile analysis.

- Be alerted when a delivery backlog builds up and monitor when the number of customer orders exceeds a defined operational threshold.

- Enable controlled order throttling during periods of sudden high demand to prevent cascading delays.

- Optimize rider route schedules based on peak-risk hours rather than average delivery demand.

- Flag and review orders exceeding 90 minutes on a weekly basis to identify repetitive operational delays.

- Enhance system hygiene by enforcing timely order closure to avoid analytical distortions caused by system artifacts.

- Track the 95th percentile delivery time continuously rather than relying on daily average metrics.

- **If these Recomedations are ignored without consideration , it might lead to SLA breach in the Future**